[CDAP-18542] Adding a periodic full compaction to LevelDBTableService.

wyzhang opened the pull request:
What:
Adding a periodic full compaction across tables managed by LevelDBTableService.
This background periodic compaction is controlled by a cdap-site configuration.
Typically this the comapction interval should be a few days (e.g. 7 days) since
it could be a heavy disk operation depending on the size of the table.

Why:
LevelDB by default compacts across level N and subset of files in level N+1, this could
potentially lead to deletion-marker-masked values in N+1 or N+2 not being discarded
after the compaction, thus degrading read performance and causing disk overhead.

fernst commented on cdap-common/src/main/java/io/cdap/cdap/common/conf/Constants.java:
```diff
@@ -1440,6 +1440,8 @@ public static String getZKQuorum(CConfiguration cConf) {
   public static final String CFG_DATA_LEVELDB_BLOCKSIZE = "data.local.storage.blocksize";
   public static final String CFG_DATA_LEVELDB_CACHESIZE = "data.local.storage.cachesize";
   public static final String CFG_DATA_LEVELDB_FSYNC = "data.local.storage.fsync";
+  public static final String CFG_DATA_LEVELDB_COMPACTION_INTERVAL_SECONDS =
```
@wyzhang would it be possible to set this in a cron expression? For some customers, it might be advisable to say "This will run every day at XX UTC"

tivv replied:
we are using quartz in app-fabric, if it's available in data fabric, it would make sense

wyzhang replied:
Let me explore this as an improvement in a follow up PR

tivv replied:
I don't think it's worth an additional PR. It's still a temporary measure, so it may not be viable to enhance it too much

wyzhang replied:
Sounds good. Will leave this as is.

tivv commented on cdap-data-fabric/src/main/java/io/cdap/cdap/data2/dataset2/lib/table/leveldb/LevelDBTableService.java:
```diff
@@ -92,6 +105,82 @@ public void setConfiguration(CConfiguration config) {
+      KeyValueDBComparator comparator = new KeyValueDBComparator();
+      // Compact all levels except the last level, since levelDB compaction will merge
+      // data in current level into next level.
+      for (int level = 0; level < DbConstants.NUM_LEVELS - 1; level++) {
```
Can we make a min and max level configurable? Doing like this will move data to the "before last" level really soon and that data will never go away. Usually compacting up to level 3 or 4 should be enough and still would leave room to compact higher level data

wyzhang replied:
Added. 

However, compaction at level X will take inputs from X and X+1, so this loop will move data all the way to the highest level. In contrast, if we only compact to level 3. This would lead to tombstoned data in level 5 and 6 never get discarded.

tivv replied:
Yes. I don't want for automated scheduled process to move data to level 6 by default as this would mean manual remediation is not available if data starts to accumulate at the highest level. On the other side, tombstones should never get to the highest level in automatic compaction just because we don't have enough data to get to those levels.

tivv replied:
Such setting would still allow to set it to compact all if we decide to, but I'd never leave it running for a prolonged period of time just because of a risk to get unmanagable big db

wyzhang replied:
Deletion-masked data could end up in highlest level without being discarded if we can compact max_level - 1. This is why we need to compact all levels (i.e. up to max_level - 1)

tivv replied:
It can end up there only in two cases:
1) We have over 1TB of data (or 10TB, not sure about this level threshold). This is not possible given storage size.
2) We pushed it there by manual compactions. 
For the second case I'd prefer to push it to the lowest level reasonable.

wyzhang replied:
This should be resolved as min and max levels are configurable now.

tivv commented on cdap-data-fabric/src/main/java/io/cdap/cdap/data2/dataset2/lib/table/leveldb/LevelDBTableService.java:
```diff
@@ -78,6 +90,7 @@ public static LevelDBTableService getInstance() {
    */
   @VisibleForTesting
   public LevelDBTableService() {
+    executor = Executors.newSingleThreadScheduledExecutor();
```
Please don't create an executor unless it's enabled. Bringing up non-daemon thread may have consequences. Also please use `org.apache.twill.common.Threads#createDaemonThreadFactory` to create daemon thread factory with a reasonable name and shut it down when service is closed

wyzhang replied:
Added a name for executor.

Looking at implementation, newSingleThreadScheduledExecutor doesn't seem to do any heavy lifting. 
Thread creation is done lazily till execution happens. 
So other then executor object memory overhead (which I suppose is negligible), seems cheap and harmless to create an executor without doing anything.

tivv replied:
For me it's just additional risk (yet small) because of additional work done when this feature is disabled. It's not about overhead, it's more about risk

wyzhang replied:
Leave as is if anyone feel strongly otherwise.

chtyim commented on cdap-data-fabric/src/main/java/io/cdap/cdap/data2/dataset2/lib/table/leveldb/LevelDBTableService.java:
```diff
@@ -92,6 +105,82 @@ public void setConfiguration(CConfiguration config) {
+
+  @VisibleForTesting
+  public void compactAll() {
+    synchronized (tables) {
```
Why need to synchronize? The tables is already a concurrent map.

wyzhang replied:
Removed. 

Original intention was to avoid scheduling compaction for the table that is being or about to be purged from disk.
But no longer needed after serializing compaction with dropping table in a single-threaded executor.

chtyim commented on cdap-data-fabric/src/main/java/io/cdap/cdap/data2/dataset2/lib/table/leveldb/LevelDBTableService.java:
```diff
@@ -92,6 +105,82 @@ public void setConfiguration(CConfiguration config) {
+    }
+    synchronized (tables) {
+      // Make sure table still exists.
+      db = tables.get(tableName);
```
This block is unnecessary. In fact, if this one does return `null`, you'll have NPE down below.

wyzhang replied:
Removed. Previously missed if == null, return.

chtyim commented on cdap-data-fabric/src/main/java/io/cdap/cdap/data2/dataset2/lib/table/leveldb/LevelDBTableService.java:
```diff
@@ -92,6 +105,82 @@ public void setConfiguration(CConfiguration config) {
+    long startMillis = System.currentTimeMillis();
+    LOG.debug("LevelDBTableService background periodic compaction on table {} started...", tableName);
+    try {
+      DbImpl dbImpl = (DbImpl) db;
```
Guard it with `instanceof` check.

wyzhang replied:
Done.

chtyim commented on cdap-data-fabric/src/main/java/io/cdap/cdap/data2/dataset2/lib/table/leveldb/LevelDBTableService.java:
```diff
@@ -92,6 +105,82 @@ public void setConfiguration(CConfiguration config) {
+    }
+
+    long startMillis = System.currentTimeMillis();
+    LOG.debug("LevelDBTableService background periodic compaction on table {} started...", tableName);
```
Don't use `...` in the message.

wyzhang replied:
Fixed.

CuriousVini commented on cdap-data-fabric/src/main/java/io/cdap/cdap/data2/dataset2/lib/table/leveldb/LevelDBTableService.java:
```diff
@@ -92,6 +105,82 @@ public void setConfiguration(CConfiguration config) {
     cacheSize = config.getLong(Constants.CFG_DATA_LEVELDB_CACHESIZE, Constants.DEFAULT_DATA_LEVELDB_CACHESIZE);
     writeOptions = new WriteOptions().sync(
       config.getBoolean(Constants.CFG_DATA_LEVELDB_FSYNC, Constants.DEFAULT_DATA_LEVELDB_FSYNC));
+    compactionInterval = Duration.ofSeconds(config.getLong(Constants.CFG_DATA_LEVELDB_COMPACTION_INTERVAL_SECONDS));
```
This config seem to be temporary? In that case should we do getLong with default value? So that we can remove it later on with correct fix?

wyzhang replied:
You are suggesting not putting it cdap-deafult?

CuriousVini replied:
The reason being 
1. This could be related to temporary fix, so there is no need to add this property to cdap-default with description which will later be deprecated.  Similar to:
https://cdap.gitee.io/docs/zh/admin-manual/appendices/cdap-site.html#deprecated-properties

2. This is kind of internal property related to level db. Mostly users will not be using it

wyzhang replied:
Done

CuriousVini commented on cdap-data-fabric/src/main/java/io/cdap/cdap/data2/dataset2/lib/table/leveldb/LevelDBTableService.java:
```diff
@@ -92,6 +105,82 @@ public void setConfiguration(CConfiguration config) {
+  }
+
+  @VisibleForTesting
+  public void compactAll() {
```
should we file a jira with correct fix for long term and add as TODO here for correct fix?

wyzhang replied:
https://cdap.atlassian.net/browse/CDAP-18546

CuriousVini replied:
Thanks for filing jira.
Usually we add it in the code with `TODO: <jira> and some description` similar to: 

https://github.com/cdapio/cdap/blob/develop/cdap-runtime-ext-remote-hadoop/src/main/java/io/cdap/cdap/runtime/spi/provisioner/remote/RemoteHadoopProvisioner.java#L139

This is helpful while refactoring code. In this case, when we implement sharding logic

wyzhang replied:
Done

CuriousVini reviewed (commented):
Thanks!

tivv commented on cdap-common/src/main/java/io/cdap/cdap/common/conf/Constants.java:
```diff
@@ -1448,6 +1454,9 @@ public static String getZKQuorum(CConfiguration cConf) {
   public static final boolean DEFAULT_DATA_LEVELDB_FSYNC = true;
+  public static final long DEFAULT_DATA_LEVELDB_COMPACTION_INTERVAL_SECONDS = 0;
+  public static final int DEFAULT_DATA_LEVELDB_COMPACTION_LEVEL_MIN = 0;
+  public static final int DEFAULT_DATA_LEVELDB_COMPACTION_LEVEL_MAX = 5;
```
I'd be more cautious by default by setting it to 4 as this is what we did for manual runs.

wyzhang replied:
SG. No strong preferences on my end. Lowered to 4.

tivv commented on cdap-data-fabric/src/main/java/io/cdap/cdap/data2/dataset2/lib/table/leveldb/LevelDBTableService.java:
```diff
@@ -78,6 +92,8 @@ public static LevelDBTableService getInstance() {
   @VisibleForTesting
   public LevelDBTableService() {
+    executor = Executors.newSingleThreadScheduledExecutor(
+      Threads.createDaemonThreadFactory("leveldb-periodic-compaction"));
```
You still want to do `executor.shutdown()` in `close`. Otherwise you'll get tons of threads created in unit tests

wyzhang replied:
I do have  shutdownNow call in close(), right?

tivv replied:
Yes, only in close

tivv reviewed (commented):
Please close executor in service `close` method. Other than that LGTM

tivv commented on cdap-common/src/main/resources/cdap-default.xml:
```diff
@@ -3116,7 +3116,7 @@
   <property>
     <name>preview.bind.port</name>
-    <value>0</value>
+    <value>11050</value>
```
Is this intentional?

wyzhang replied:
Unintentional. Reverted.

tivv commented on cdap-data-fabric/src/main/java/io/cdap/cdap/data2/dataset2/lib/table/leveldb/LevelDBTableService.java:
```diff
@@ -90,8 +106,88 @@ public void setConfiguration(CConfiguration config) {
+    compactionLevelMax = config.getInt(Constants.CFG_DATA_LEVELDB_COMPACTION_LEVEL_MAX,
+                                       Constants.DEFAULT_DATA_LEVELDB_COMPACTION_LEVEL_MAX);
+    if (compactionInterval.getSeconds() > 0) {
+      executor.shutdownNow();
```
Why are you closing an executor here?

tivv replied:
It does not say it in docs for `shutdownNow`, but `shutdown` says it won't accept any more tasks. I suspect `shutdownNow` is same

wyzhang replied:
ShutdownNow does a super set of what shutdown does. 

shtudown -> SHUTDOWN: Don't accept new tasks, but process queued tasks 
shutdownNow -> STOP: Don't accept new tasks, don't process queued tasks, and interrupt in-progress tasks TIDYING: All

tivv commented on cdap-data-fabric/src/main/java/io/cdap/cdap/data2/dataset2/lib/table/leveldb/LevelDBTableService.java:
```diff
@@ -90,8 +106,81 @@ public void setConfiguration(CConfiguration config) {
+                                       Constants.DEFAULT_DATA_LEVELDB_COMPACTION_LEVEL_MIN);
+    compactionLevelMax = config.getInt(Constants.CFG_DATA_LEVELDB_COMPACTION_LEVEL_MAX,
+                                       Constants.DEFAULT_DATA_LEVELDB_COMPACTION_LEVEL_MAX);
+    executor.shutdownNow();
```
Don't shutdown it in the setConfiguration at all, only in close

wyzhang replied:
Original intention to stop the scheduled task. Cannot be achieved via shutdown as you mentioned. 

Updated. PTAL

chtyim commented on cdap-data-fabric/src/main/java/io/cdap/cdap/data2/dataset2/lib/table/leveldb/LevelDBTableService.java:
```diff
@@ -58,12 +68,18 @@
   private boolean compressionEnabled;
   private int blockSize;
   private long cacheSize;
+  private Duration compactionInterval;
```
All these can be final fields?

wyzhang replied:
Yes. They can. But needs a small refactoring touching a few files and tests. Will do in a follow-up PR, so this PR can be small and easy to cherrypick.

chtyim commented on cdap-data-fabric/src/main/java/io/cdap/cdap/data2/dataset2/lib/table/leveldb/LevelDBTableService.java:
```diff
@@ -90,8 +108,84 @@ public void setConfiguration(CConfiguration config) {
+  public void compactAll() {
+    // TODO CDAP-18546: deprecate compaction in favor of using sharding for efficient recycling range deleted rows.
+    for (Map.Entry<String, DB> entry : tables.entrySet()) {
+      compact(entry.getKey());
```
Why not just pass the `entry.value()` in such that you don't need to get and check for null in the compact method.

wyzhang replied:
Passing a name allows us to print db name in logs, thus done that way.

Doing another lookup is a tiny optimization to avoid attempting compaction entirely in case table gets dropped between the table lookup and actual compaction execution.

chtyim commented on cdap-data-fabric/src/main/java/io/cdap/cdap/data2/dataset2/lib/table/leveldb/LevelDBTableService.java:
```diff
@@ -90,8 +108,84 @@ public void setConfiguration(CConfiguration config) {
+    LOG.debug("LevelDBTableService background periodic compaction on table {} started.", tableName);
+    try {
+      if (!(db instanceof DbImpl)) {
+        throw new ClassCastException(String.format("DB for table %s is not an object of DbImpl ", tableName));
```
Better just log and skip it rather than throwing exception.

wyzhang replied:
Done

chtyim commented on cdap-data-fabric/src/main/java/io/cdap/cdap/data2/dataset2/lib/table/leveldb/LevelDBTableService.java:
```diff
@@ -78,6 +94,8 @@ public static LevelDBTableService getInstance() {
    */
   @VisibleForTesting
   public LevelDBTableService() {
+    executor = Executors.newSingleThreadScheduledExecutor(
```
New it the last such that if there is any exception thrown, the execution won't get created

wyzhang replied:
Currently there is no other news in the constructor. I assume you are referring to refactoring variables to final and init them in the constructor, then make sure the executor gets init'ed the last, right? Will do that when making refactoring change in a follow  PR.

wyzhang committed:
[CDAP-18542] Adding a periodic full compaction to LevelDBTableService.

What:
Adding a periodic full compaction across tables managed by LevelDBTableService.
This background periodic compaction is controlled by a cdap-site configuration.
Typically this the comapction interval should be a few days (e.g. 7 days) since
it could be a heavy disk operation depending on the size of the table.

Why:
LevelDB by default compacts across level N and subset of files in level N+1, this could
potentially lead to deletion-marker-masked values in N+1 or N+2 not being discarded
after the compaction, thus degrading read performance and causing disk overhead.

wyzhang commented:
@chtyim  I am going to merge this PR and then cherrypick to unblock building patch release. Please leave comments if you have any, will address in a separate PR
